029-across-step/proposal.md

```diff
@@ -0,0 +1,321 @@
+full matrix completes.
+
+`attempts` binds to the inner step because it doesn't seem to make a whole lot
+of sense to retry the *entire matrix* because one step failed. The individual
+steps should be retried instead.
+
+`timeout` binds to the inner step because otherwise a `max_in_flight` could
+cause the timeout to be exceeded before some steps even get a chance to run.
+
+```yaml
+task: unit
+timeout: 1h # interrupt the task after 1 hour
+attempts: 3 # attempt the task 3 times
+across:
+- var: go_version
+  values: [1.12, 1.13]
+on_failure: # do something after all steps complete and at least one failed
+```
+
+To apply `ensure` and `on_*` hooks to the nested step, rather than the `across`
+step modifier, the `do:` step may be utilized:
+
+```yaml
+do:
+- task: unit
+  on_failure: # runs after each individual step completes and fails
+across:
+- var: go_version
+  values: [1.12, 1.13]
+on_failure: # runs after all steps complete and at least one failed
+```
+
+This can be rewritten in a slightly more readable syntax by placing the `do:`
+below the `across:`:
+
+```yaml
+across:
+- var: go_version
+  values: [1.12, 1.13]
+do:
+- task: unit
+  on_failure: # runs after each individual step completes and fails
+on_failure: # runs after all steps complete and at least one failed
+```
+
+### Failing fast
+
+When a step within the matrix fails, the `across` step will continue to run the
+remaining steps in the matrix. However, the `across` step itself will still
+fail after all steps have completed.
+
+With `fail_fast: true` applied to the `across` step, execution will halt on the
+first failure, and any currently running steps (via `max_in_flight`) will be
+interrupted:
+
+```yaml
+task: unit
+across:
+- var: go_version
+  values: [1.12, 1.13]
+fail_fast: true
+```
+
+Note: this is the first time a step *modifier* has had additional sibling
+fields. In the event of a field conflict, the `do:` step may be utilized as a
+work-around.
+
+### Var scoping and shadowing
+
+The inner step will be run with a local var scope that inherits from the outer
+scope and is initialized with the across step's var values.
+
+Given that it runs with its own scope, it follows that the vars in the local
+scope, along with any vars newly set within the scope, are not accessible
+outside of the `across` step.
+
+When a var set by the `across` step shadows outer vars, a warning will be
+printed.
+
+Example:
+
+```yaml
+plan:
+- get_var: foo
+  trigger: true
+- run: print
+  type: debug
+  params: {value: ((.:foo))}
+- across:
+  - var: foo
+    values: [one, two]
+  do:
+  - run: print
+    type: debug
+    params: {value: ((.:foo))}
+- run: print
+  type: debug
+  params: {value: ((.:foo))}
+```
+
+Assuming the `get_var` step produces a value of `zero`, this build plan should
+result in the following output:
+
+```
+zero
+WARNING: across step shadows local var 'foo'
+one
+two
+zero
+```
+
+
+## Open Questions
+
+* n/a
```

Comment:
@JohannesRudolph Sorry, completely forgot to address your point! (Raising this as a review comment so the replies can be threaded.)

Would it be sufficient to allow a static list of combinations to skip to be provided?

```yaml
across:
- var: go
  values: [1.16, 1.15, 1.8]
- var: platform
  values: [darwin, linux, windows]
except:
- platform: windows
  go: 1.18
```

This would work similarly to [excluding jobs in Travis CI](https://docs.travis-ci.com/user/build-matrix/#excluding-jobs), i.e. a subset of vars may be specified, and all variations with those vars will be skipped.

Reply:
Related thought: it would be kind of cool if we could `except` combinations based on a relationship between variables. e.g. if you want to test various upgrade paths, you only want to test combinations of `from` and `to` where `to > from`. I have no idea how this would work - it might just be a use case for var sources rather than `except`